Improve `ImmutableArrayExtensions.SequenceEqual`

prozolic opened the pull request:
# Summary

This PR improves the performance of ImmutableArrayExtensions.SequenceEqual method through optimized implementation strategies:

Key Performance Improvements:
 - Array: ~90% faster (68.885ns → 6.856ns)
 - List: ~91% faster (85.014ns → 7.676ns)
 - IList: ~54% faster (125.897ns → 57.836ns)
 - ICollection: No statistically significant performance difference
 - IEnumerable: Uses existing implementation for this case because the existing implementation had better performance and didn't allocate a second enumerator
  
# Implementation Approach

The implementation was refined based on review feedback from @neon-sunset and @xtqqczze to achieve optimal performance while balancing complexity:

1. Fast path for common collections: Detects when items implement `ICollection<T>` and delegates to the highly optimized `Enumerable.SequenceEqual` using the underlying array
2. Compatibility preservation: Falls back to existing implementation for other `IEnumerable<T>` types to maintain behavioral consistency

# Benchmark (https://github.com/dotnet/performance/pull/4915)

Full benchmark results: [MihuBot/runtime-utils#1565](https://github.com/MihuBot/runtime-utils/issues/1565) (X64)

| Method        | Toolchain | input       | Mean       | Error      | Ratio | Allocated | Alloc Ratio |
|-------------- |---------- |------------ |-----------:|-----------:|------:|----------:|------------:|
| SequenceEqual | Main      | Array       |  68.885 ns |  0.1938 ns |  1.00 | -         | NA          |
| SequenceEqual | PR        | Array       |   6.856 ns |  0.2262 ns |  0.10 | -         | NA          |
|               |           |             |            |            |       |           |             |
| SequenceEqual | Main      | ICollection | 120.542 ns |  7.4864 ns |  1.00 | 32 B      | 1.00        |
| SequenceEqual | PR        | ICollection | 114.458 ns |  1.2362 ns |  0.95 | 32 B      | 1.00        |
|               |           |             |            |            |       |           |             |
| SequenceEqual | Main      | IEnumerable | 121.923 ns |  7.7103 ns |  1.00 | 32 B      | 1.00        |
| SequenceEqual | PR        | IEnumerable | 118.370 ns |  1.4443 ns |  0.98 | 32 B      | 1.00        |
|               |           |             |            |            |       |           |             |
| SequenceEqual | Main      | IList       | 125.897 ns |  9.9231 ns |  1.01 | 32 B      | 1.00        |
| SequenceEqual | PR        | IList       |  57.836 ns |  0.9568 ns |  0.46 | -         | 0.00        |
|               |           |             |            |            |       |           |             |
| SequenceEqual | Main      | List        |  85.014 ns |  1.7593 ns |  1.00 | -         | NA          |
| SequenceEqual | PR        | List        |   7.676 ns |  0.2719 ns |  0.09 | -         | NA          |

prozolic committed:
Improve ImmutableArrayExtensions.SequenceEqual

prozolic commented:
@dotnet-policy-service agree

xtqqczze commented on src/libraries/System.Collections.Immutable/src/System/Linq/ImmutableArrayExtensions.cs:
```diff
@@ -1,4 +1,4 @@
-// Licensed to the .NET Foundation under one or more agreements.
+﻿// Licensed to the .NET Foundation under one or more agreements.
```
Please remove the Unicode BOM (\uFEFF) from this file.

prozolic committed:
Remove BOM

PranavSenthilnathan commented on src/libraries/System.Collections.Immutable/src/System/Linq/ImmutableArrayExtensions.cs:
```diff
@@ -189,9 +189,32 @@ public static bool SequenceEqual<TDerived, TBase>(this ImmutableArray<TBase> imm
+                    return false;
+                }
+
+                if (items is IList<TDerived> itemList)
+                {
+                    comparer ??= EqualityComparer<TBase>.Default;
+
+                    for (i = 0; i < count; i++)
+                    {
+                        if (!comparer.Equals(immutableArray[i], itemList[i]))
+                        {
+                            return false;
+                        }
+                    }
+
+                    return true;
+                }
```
I think the JIT should be able to de-abstract the foreach loop below instead of us doing it here. How much perf benefit is there with/without this?

prozolic replied:
@PranavSenthilnathan 
I see, I had overlooked that perspective. I performed the performance comparison between manual optimization and foreach loops. The results show that manual optimization runs significantly faster and also reduces memory allocations.

Results (Size 1000):
- List: Manual 549ns/0B vs Foreach 1,439ns/40B → **2.6x faster, zero allocations**
- Array: Manual 382ns/0B vs Foreach 1,531ns/32B → **4x faster, zero allocations**

<details>
<summary>Full benchmark results</summary>

```
| Method                      | Size | Mean          | Error      | StdDev     | Gen0   | Allocated |
|---------------------------- |----- |--------------:|-----------:|-----------:|-------:|----------:|
| SequenceEqual_Sequence      | 10   |    28.3406 ns |  0.7532 ns |  0.8674 ns | 0.0048 |      40 B |
| SequenceEqual_Array         | 10   |     6.0947 ns |  0.0331 ns |  0.0293 ns |      - |         - |
| SequenceEqual_List          | 10   |     8.9099 ns |  0.1671 ns |  0.1563 ns |      - |         - |
| SequenceEqual_Sequence2     | 10   |    35.9718 ns |  0.8827 ns |  0.9445 ns | 0.0047 |      40 B |
| SequenceEqual_Array2        | 10   |     1.2575 ns |  0.0551 ns |  0.0635 ns |      - |         - |
| SequenceEqual_List2         | 10   |     1.4464 ns |  0.0357 ns |  0.0316 ns |      - |         - |
| Jit_SequenceEqual_Sequence  | 10   |    41.6518 ns |  0.5336 ns |  0.4456 ns | 0.0047 |      40 B |
| Jit_SequenceEqual_Array     | 10   |    23.1111 ns |  0.6083 ns |  0.6761 ns | 0.0038 |      32 B |
| Jit_SequenceEqual_List      | 10   |    24.8457 ns |  0.5393 ns |  0.5995 ns | 0.0047 |      40 B |
| Jit_SequenceEqual_Sequence2 | 10   |    47.1824 ns |  0.6800 ns |  0.5678 ns | 0.0047 |      40 B |
| Jit_SequenceEqual_Array2    | 10   |     1.6589 ns |  0.0474 ns |  0.0465 ns |      - |         - |
| Jit_SequenceEqual_List2     | 10   |     1.1552 ns |  0.0228 ns |  0.0213 ns |      - |         - |
| SequenceEqual_Sequence      | 100  |   194.1603 ns |  0.9953 ns |  0.8823 ns | 0.0047 |      40 B |
| SequenceEqual_Array         | 100  |    46.3683 ns |  0.8452 ns |  0.7058 ns |      - |         - |
| SequenceEqual_List          | 100  |    61.2435 ns |  0.9109 ns |  0.7112 ns |      - |         - |
| SequenceEqual_Sequence2     | 100  |   247.6734 ns |  5.6015 ns |  6.2260 ns | 0.0039 |      40 B |
| SequenceEqual_Array2        | 100  |     1.1905 ns |  0.0306 ns |  0.0287 ns |      - |         - |
| SequenceEqual_List2         | 100  |     1.4367 ns |  0.0313 ns |  0.0293 ns |      - |         - |
| Jit_SequenceEqual_Sequence  | 100  |   291.8718 ns |  4.1452 ns |  3.6746 ns | 0.0048 |      40 B |
| Jit_SequenceEqual_Array     | 100  |   165.9619 ns |  1.5901 ns |  1.3278 ns | 0.0033 |      32 B |
| Jit_SequenceEqual_List      | 100  |   156.7944 ns |  0.6257 ns |  0.5853 ns | 0.0044 |      40 B |
| Jit_SequenceEqual_Sequence2 | 100  |   353.6850 ns |  4.9917 ns |  4.1683 ns | 0.0042 |      40 B |
| Jit_SequenceEqual_Array2    | 100  |     1.0040 ns |  0.0345 ns |  0.0322 ns |      - |         - |
| Jit_SequenceEqual_List2     | 100  |     0.9756 ns |  0.0064 ns |  0.0057 ns |      - |         - |
| SequenceEqual_Sequence      | 1000 | 1,849.0704 ns | 31.1836 ns | 29.1692 ns |      - |      40 B |
| SequenceEqual_Array         | 1000 |   382.4244 ns |  8.8805 ns |  9.5020 ns |      - |         - |
| SequenceEqual_List          | 1000 |   549.2625 ns |  4.1302 ns |  3.4489 ns |      - |         - |
| SequenceEqual_Sequence2     | 1000 | 1,851.6025 ns | 30.2908 ns | 28.3340 ns |      - |      40 B |
| SequenceEqual_Array2        | 1000 |     1.2066 ns |  0.0092 ns |  0.0082 ns |      - |         - |
| SequenceEqual_List2         | 1000 |     2.2791 ns |  0.1321 ns |  0.1469 ns |      - |         - |
| Jit_SequenceEqual_Sequence  | 1000 | 2,790.9678 ns | 60.2994 ns | 67.0226 ns |      - |      40 B |
| Jit_SequenceEqual_Array     | 1000 | 1,530.9895 ns | 29.2066 ns | 24.3888 ns |      - |      32 B |
| Jit_SequenceEqual_List      | 1000 | 1,438.9016 ns | 24.0224 ns | 22.4705 ns |      - |      40 B |
| Jit_SequenceEqual_Sequence2 | 1000 | 2,873.5082 ns | 73.5083 ns | 81.7043 ns |      - |      40 B |
| Jit_SequenceEqual_Array2    | 1000 |     1.0051 ns |  0.0294 ns |  0.0245 ns |      - |         - |
| Jit_SequenceEqual_List2     | 1000 |     1.0230 ns |  0.0418 ns |  0.0464 ns |      - |         - |
```

```csharp
        public static bool SequenceEqual<TDerived, TBase>(this ImmutableArray<TBase> immutableArray, IEnumerable<TDerived> items, IEqualityComparer<TBase>? comparer = null) where TDerived : TBase
        {
            Requires.NotNull(items, nameof(items));

            int i = 0;
            if (items.TryGetCount(out var count))
            {
                if (immutableArray.Length != count)
                {
                    return false;
                }

                if (items is IList<TDerived> itemList)
                {
                    comparer ??= EqualityComparer<TBase>.Default;

                    for (i = 0; i < count; i++)
                    {
                        if (!comparer.Equals(immutableArray[i], itemList[i]))
                        {
                            return false;
                        }
                    }

                    return true;
                }
            }

            comparer ??= EqualityComparer<TBase>.Default;

            int n = immutableArray.Length;
            foreach (TDerived item in items)
            {
                if (i == n)
                {
                    return false;
                }

                if (!comparer.Equals(immutableArray[i], item))
                {
                    return false;
                }

                i++;
            }

            return i == n;
        }

        public static bool SequenceEqual_Jit<TDerived, TBase>(this ImmutableArray<TBase> immutableArray, IEnumerable<TDerived> items, IEqualityComparer<TBase>? comparer = null) where TDerived : TBase
        {
            Requires.NotNull(items, nameof(items));

            int i = 0;
            if (items.TryGetCount(out var count))
            {
                if (immutableArray.Length != count)
                {
                    return false;
                }

                if (items is IList<TDerived> itemList)
                {
                    comparer ??= EqualityComparer<TBase>.Default;

                    foreach (TDerived item in itemList)
                    {
                        if (!comparer.Equals(immutableArray[i], item))
                        {
                            return false;
                        }
                        i++;
                    }
                    return true;
                }
            }

            comparer ??= EqualityComparer<TBase>.Default;

            int n = immutableArray.Length;
            foreach (TDerived item in items)
            {
                if (i == n)
                {
                    return false;
                }

                if (!comparer.Equals(immutableArray[i], item))
                {
                    return false;
                }

                i++;
            }

            return i == n;
        }
```

```csharp
    public class ImmutableArrayExtensionsBenchmark
    {
        private ImmutableArray<int> _immutableArray;
        private IEnumerable<int> _sequence;
        private IEnumerable<int> _array;
        private IEnumerable<int> _list;

        private IEnumerable<int> _sequence2;
        private IEnumerable<int> _array2;
        private IEnumerable<int> _list2;

        [Params(10, 100, 1000)]
        public int Size { get; set; }

        [GlobalSetup]
        public void GlobalSetup()
        {
            _immutableArray = Enumerable.Range(0, Size).ToImmutableArray();
            _sequence = Sequence(Enumerable.Range(0, Size));
            _array = Enumerable.Range(0, Size).ToArray();
            _list = Enumerable.Range(0, Size).ToList();

            _sequence2 = Sequence(Enumerable.Range(0, Size + 1));
            _array2 = Enumerable.Range(0, Size + 1).ToArray();
            _list2 = Enumerable.Range(0, Size + 1).ToList();
        }

        private static IEnumerable<TItem> Sequence<TItem>(IEnumerable<TItem> source)
        {
            foreach (var item in source)
                yield return item;
        }

        [Benchmark]
        public bool SequenceEqual_Sequence()
        {
            return _immutableArray.SequenceEqual(_sequence);
        }

        [Benchmark]
        public bool SequenceEqual_Array()
        {
            return _immutableArray.SequenceEqual(_array);
        }

        [Benchmark]
        public bool SequenceEqual_List()
        {
            return _immutableArray.SequenceEqual(_list);
        }

        [Benchmark]
        public bool SequenceEqual_Sequence2()
        {
            return _immutableArray.SequenceEqual(_sequence2);
        }

        [Benchmark]
        public bool SequenceEqual_Array2()
        {
            return _immutableArray.SequenceEqual(_array2);
        }

        [Benchmark]
        public bool SequenceEqual_List2()
        {
            return _immutableArray.SequenceEqual(_list2);
        }


        [Benchmark]
        public bool Jit_SequenceEqual_Sequence()
        {
            return _immutableArray.SequenceEqual_Jit(_sequence);
        }

        [Benchmark]
        public bool Jit_SequenceEqual_Array()
        {
            return _immutableArray.SequenceEqual_Jit(_array);
        }

        [Benchmark]
        public bool Jit_SequenceEqual_List()
        {
            return _immutableArray.SequenceEqual_Jit(_list);
        }

        [Benchmark]
        public bool Jit_SequenceEqual_Sequence2()
        {
            return _immutableArray.SequenceEqual_Jit(_sequence2);
        }

        [Benchmark]
        public bool Jit_SequenceEqual_Array2()
        {
            return _immutableArray.SequenceEqual_Jit(_array2);
        }

        [Benchmark]
        public bool Jit_SequenceEqual_List2()
        {
            return _immutableArray.SequenceEqual_Jit(_list2);
        }
    }
```
</details>

PranavSenthilnathan commented on src/libraries/System.Collections.Immutable/src/System/Linq/ImmutableArrayExtensions.cs:
```diff
@@ -189,9 +189,32 @@ public static bool SequenceEqual<TDerived, TBase>(this ImmutableArray<TBase> imm
             Requires.NotNull(items, nameof(items));
 
+            int i = 0;
+            if (items.TryGetCount(out var count))
```
```suggestion
            if (items.TryGetCount(out int count))
```

neon-sunset commented:
Perhaps a simpler (and faster) way to do this is to just call `ImmutableCollectionsMarshal.AsArray` and delegate to already optimized `Enumerable.SequenceEqual` by using the underlying array instead?

prozolic committed:
Update src/libraries/System.Collections.Immutable/src/System/Linq/ImmutableArrayExtensions.cs

Co-authored-by: Pranav Senthilnathan <pranav.senthilnathan@live.com>

prozolic commented:
I'm not sure if it can be applied to `ImmutableArrayExtensions.SequenceEqual`, but I also think that is the simplest and fastest approach.

xtqqczze commented:
> I'm not sure if it can be applied to `ImmutableArrayExtensions.SequenceEqual`, but I also think that is the simplest and fastest approach.

This should work:

```cs
immutableArray.ThrowNullRefIfNotInitialized();
return immutableArray.array.SequenceEqual((IEnumerable<TBase>)items, comparer);
```

prozolic committed:
Replace manual implementation with Enumerable.SequenceEqual

xtqqczze commented:
> The performance is also better than what I originally proposed, so I've changed the implementation to this approach.

@prozolic I see a few regressions. Could you collect benchmark data for https://github.com/xtqqczze/dotnet-runtime/tree/ImmutableArrayExtensions.SequenceEqual?. This approach will not be worth the additional complexity, however.

prozolic commented:
Sorry. Indeed, there were memory allocation issues and regressions in some cases. I conducted performance measurements comparing different implementations of SequenceEqual. The results include the following implementation variants:

- SequenceEqual_Original: Current existing implementation
- SequenceEqual_1: My initially proposed implementation
- SequenceEqual_2: @xtqqczze's implementation approach
- SequenceEqual_3: Implementation using `Enumerable.SequenceEqual`

<details>
<summary>Full benchmark results</summary>

```
| Method                          | Size | Mean         | Error      | StdDev     | Median       | Min          | Max          | Gen0   | Allocated |
|-------------------------------- |----- |-------------:|-----------:|-----------:|-------------:|-------------:|-------------:|-------:|----------:|
| SequenceEqual_Original_Sequence | 10   |    29.367 ns |  0.6282 ns |  0.6982 ns |    29.062 ns |    28.659 ns |    30.968 ns | 0.0047 |      40 B |
| SequenceEqual_Original_Array    | 10   |     7.832 ns |  0.2126 ns |  0.2448 ns |     7.759 ns |     7.563 ns |     8.346 ns |      - |         - |
| SequenceEqual_Original_List     | 10   |    10.526 ns |  0.1244 ns |  0.1103 ns |    10.479 ns |    10.440 ns |    10.755 ns |      - |         - |
| SequenceEqual_1_Sequence        | 10   |    29.635 ns |  0.7748 ns |  0.8290 ns |    29.291 ns |    28.906 ns |    31.585 ns | 0.0047 |      40 B |
| SequenceEqual_1_Array           | 10   |     6.404 ns |  0.1304 ns |  0.1219 ns |     6.357 ns |     6.278 ns |     6.694 ns |      - |         - |
| SequenceEqual_1_List            | 10   |     8.755 ns |  0.1162 ns |  0.0970 ns |     8.706 ns |     8.612 ns |     8.915 ns |      - |         - |
| SequenceEqual_2_Sequence        | 10   |    32.641 ns |  0.5341 ns |  0.4735 ns |    32.537 ns |    32.014 ns |    33.533 ns | 0.0047 |      40 B |
| SequenceEqual_2_Array           | 10   |     3.557 ns |  0.0850 ns |  0.0710 ns |     3.515 ns |     3.497 ns |     3.709 ns |      - |         - |
| SequenceEqual_2_List            | 10   |     2.711 ns |  0.0300 ns |  0.0266 ns |     2.702 ns |     2.687 ns |     2.772 ns |      - |         - |
| SequenceEqual_3_Sequence        | 10   |    36.971 ns |  0.5349 ns |  0.4742 ns |    36.810 ns |    36.374 ns |    38.207 ns | 0.0085 |      72 B |
| SequenceEqual_3_Array           | 10   |     4.584 ns |  0.0485 ns |  0.0430 ns |     4.566 ns |     4.545 ns |     4.684 ns |      - |         - |
| SequenceEqual_3_List            | 10   |     3.958 ns |  0.0137 ns |  0.0107 ns |     3.957 ns |     3.942 ns |     3.979 ns |      - |         - |
| SequenceEqual_Original_Sequence | 100  |   196.088 ns |  1.4928 ns |  1.1655 ns |   196.009 ns |   194.441 ns |   198.118 ns | 0.0047 |      40 B |
| SequenceEqual_Original_Array    | 100  |    64.351 ns |  0.4256 ns |  0.3554 ns |    64.235 ns |    64.013 ns |    65.284 ns |      - |         - |
| SequenceEqual_Original_List     | 100  |    84.848 ns |  1.1737 ns |  1.0979 ns |    84.272 ns |    83.994 ns |    87.165 ns |      - |         - |
| SequenceEqual_1_Sequence        | 100  |   196.206 ns |  0.8333 ns |  0.7387 ns |   196.087 ns |   195.027 ns |   197.763 ns | 0.0040 |      40 B |
| SequenceEqual_1_Array           | 100  |    46.198 ns |  0.1243 ns |  0.1038 ns |    46.171 ns |    46.036 ns |    46.408 ns |      - |         - |
| SequenceEqual_1_List            | 100  |    60.603 ns |  0.1870 ns |  0.1460 ns |    60.550 ns |    60.443 ns |    60.870 ns |      - |         - |
| SequenceEqual_2_Sequence        | 100  |   195.790 ns |  0.8383 ns |  0.7000 ns |   195.679 ns |   194.953 ns |   197.149 ns | 0.0047 |      40 B |
| SequenceEqual_2_Array           | 100  |     6.371 ns |  0.1426 ns |  0.1264 ns |     6.353 ns |     6.161 ns |     6.576 ns |      - |         - |
| SequenceEqual_2_List            | 100  |     5.360 ns |  0.0521 ns |  0.0435 ns |     5.342 ns |     5.324 ns |     5.482 ns |      - |         - |
| SequenceEqual_3_Sequence        | 100  |   243.019 ns |  1.9271 ns |  1.7083 ns |   242.432 ns |   241.329 ns |   247.212 ns | 0.0078 |      72 B |
| SequenceEqual_3_Array           | 100  |     6.411 ns |  0.0080 ns |  0.0063 ns |     6.411 ns |     6.400 ns |     6.422 ns |      - |         - |
| SequenceEqual_3_List            | 100  |     6.652 ns |  0.0175 ns |  0.0164 ns |     6.653 ns |     6.629 ns |     6.686 ns |      - |         - |
| SequenceEqual_Original_Sequence | 1000 | 1,818.861 ns | 10.4188 ns |  8.1343 ns | 1,817.715 ns | 1,810.100 ns | 1,835.355 ns |      - |      40 B |
| SequenceEqual_Original_Array    | 1000 |   645.510 ns |  1.0769 ns |  0.9547 ns |   645.590 ns |   643.858 ns |   646.983 ns |      - |         - |
| SequenceEqual_Original_List     | 1000 | 1,295.138 ns |  4.3377 ns |  3.3866 ns | 1,294.743 ns | 1,292.376 ns | 1,305.142 ns |      - |      40 B |
| SequenceEqual_1_Sequence        | 1000 | 1,594.272 ns | 17.2476 ns | 15.2896 ns | 1,591.854 ns | 1,574.525 ns | 1,627.906 ns |      - |      40 B |
| SequenceEqual_1_Array           | 1000 |   374.425 ns |  1.1887 ns |  0.9926 ns |   374.156 ns |   373.293 ns |   376.571 ns |      - |         - |
| SequenceEqual_1_List            | 1000 |   548.014 ns |  1.5888 ns |  1.4084 ns |   547.626 ns |   546.568 ns |   551.367 ns |      - |         - |
| SequenceEqual_2_Sequence        | 1000 | 1,819.163 ns | 11.5486 ns |  9.6436 ns | 1,814.556 ns | 1,809.856 ns | 1,837.792 ns |      - |      40 B |
| SequenceEqual_2_Array           | 1000 |    34.652 ns |  0.0900 ns |  0.0752 ns |    34.639 ns |    34.563 ns |    34.813 ns |      - |         - |
| SequenceEqual_2_List            | 1000 |    34.752 ns |  0.0777 ns |  0.0606 ns |    34.743 ns |    34.675 ns |    34.846 ns |      - |         - |
| SequenceEqual_3_Sequence        | 1000 | 2,195.755 ns | 12.7553 ns | 10.6512 ns | 2,194.216 ns | 2,184.508 ns | 2,224.043 ns |      - |      72 B |
| SequenceEqual_3_Array           | 1000 |    34.630 ns |  0.3921 ns |  0.3476 ns |    34.461 ns |    34.343 ns |    35.497 ns |      - |         - |
| SequenceEqual_3_List            | 1000 |    35.973 ns |  0.5714 ns |  0.4771 ns |    35.882 ns |    35.506 ns |    37.247 ns |      - |         - |
```

```csharp
#if NET6_0_OR_GREATER

        public static bool SequenceEqual_2<TDerived, TBase>(this ImmutableArray<TBase> immutableArray, IEnumerable<TDerived> items, IEqualityComparer<TBase>? comparer = null) where TDerived : TBase
        {
            immutableArray.ThrowNullRefIfNotInitialized();
            Requires.NotNull(items, nameof(items));

            int i = 0;

            if (items is ICollection<TDerived> itemsCol)
            {
                if (((IEnumerable<TBase>)items).TryGetSpan(out ReadOnlySpan<TBase> itemsSpan))
                {
                    return immutableArray.array!.SequenceEqual(itemsSpan, comparer);
                }

                if (immutableArray.array!.Length != itemsCol.Count)
                {
                    return false;
                }

                if (itemsCol is IList<TDerived> itemsList)
                {
                    comparer ??= EqualityComparer<TBase>.Default;

                    int count = immutableArray.array!.Length;
                    for (i = 0; i < count; i++)
                    {
                        if (!comparer.Equals(immutableArray.array![i], itemsList[i]))
                        {
                            return false;
                        }
                    }

                    return true;
                }
            }

            comparer ??= EqualityComparer<TBase>.Default;

            int n = immutableArray.array!.Length;
            foreach (TDerived item in items)
            {
                if (i == n)
                {
                    return false;
                }

                if (!comparer.Equals(immutableArray.array![i], item))
                {
                    return false;
                }

                i++;
            }

            return i == n;
        }

        [MethodImpl(MethodImplOptions.AggressiveInlining)] // fast type checks that don't add a lot of overhead
        private static bool TryGetSpan<TSource>(this IEnumerable<TSource> source, out ReadOnlySpan<TSource> span)
        {
            // Use `GetType() == typeof(...)` rather than `is` to avoid cast helpers.  This is measurably cheaper
            // but does mean we could end up missing some rare cases where we could get a span but don't (e.g. a uint[]
            // masquerading as an int[]).  That's an acceptable tradeoff.  The Unsafe usage is only after we've
            // validated the exact type; this could be changed to a cast in the future if the JIT starts to recognize it.
            // We only pay the comparison/branching costs here for super common types we expect to be used frequently
            // with LINQ methods.

            bool result = true;
            if (source.GetType() == typeof(TSource[]))
            {
                span = Unsafe.As<TSource[]>(source);
            }
            else if (source.GetType() == typeof(List<TSource>))
            {
                span = CollectionsMarshal.AsSpan(Unsafe.As<List<TSource>>(source));
            }
            else
            {
                span = default;
                result = false;
            }

            return result;
        }

#endif
```

</details>

xtqqczze commented:
Something like:

```cs
        {
            if (items is ICollection<TBase> itemsCol)
            {
                immutableArray.ThrowNullRefIfNotInitialized();
                return immutableArray.array!.SequenceEqual(itemsCol, comparer);
            }
            
            return Impl(immutableArray, items, comparer);
        }
```

and then put argument validation and existing implementation in static local function `Impl`.

xtqqczze commented:
@prozolic You don't need the `comparer = null` in the `Impl` signature.

Could you also provide benchmark data again?

prozolic commented:
@xtqqczze Here are the benchmark results comparing with the current existing implementation (SequenceEqual_Original):

```
| Method                          | Size | Mean         | Error      | StdDev     | Median       | Min          | Max          | Gen0   | Allocated |
|-------------------------------- |----- |-------------:|-----------:|-----------:|-------------:|-------------:|-------------:|-------:|----------:|
| SequenceEqual_Original_Sequence | 10   |    32.759 ns |  2.9278 ns |  3.1327 ns |    32.048 ns |    29.604 ns |    40.901 ns | 0.0047 |      40 B |
| SequenceEqual_Original_Array    | 10   |     7.886 ns |  0.1893 ns |  0.2104 ns |     7.818 ns |     7.608 ns |     8.256 ns |      - |         - |
| SequenceEqual_Original_List     | 10   |    10.709 ns |  0.2367 ns |  0.2325 ns |    10.699 ns |    10.438 ns |    11.160 ns |      - |         - |
| SequenceEqual_Sequence          | 10   |    29.727 ns |  0.6673 ns |  0.7140 ns |    29.684 ns |    28.867 ns |    31.021 ns | 0.0047 |      40 B |
| SequenceEqual_Array             | 10   |     3.704 ns |  0.5038 ns |  0.5802 ns |     3.477 ns |     3.007 ns |     4.748 ns |      - |         - |
| SequenceEqual_List              | 10   |     3.410 ns |  0.1944 ns |  0.2080 ns |     3.347 ns |     3.196 ns |     3.946 ns |      - |         - |
| SequenceEqual_Original_Sequence | 100  |   197.313 ns |  2.6539 ns |  2.3526 ns |   196.760 ns |   193.830 ns |   201.666 ns | 0.0043 |      40 B |
| SequenceEqual_Original_Array    | 100  |    66.773 ns |  2.8033 ns |  2.9995 ns |    65.365 ns |    64.431 ns |    73.763 ns |      - |         - |
| SequenceEqual_Original_List     | 100  |    92.793 ns |  5.7551 ns |  6.6276 ns |    91.450 ns |    84.552 ns |   109.164 ns |      - |         - |
| SequenceEqual_Sequence          | 100  |   201.575 ns |  4.7906 ns |  5.1259 ns |   200.430 ns |   195.562 ns |   215.626 ns | 0.0047 |      40 B |
| SequenceEqual_Array             | 100  |     6.953 ns |  0.0702 ns |  0.0548 ns |     6.950 ns |     6.889 ns |     7.069 ns |      - |         - |
| SequenceEqual_List              | 100  |     7.090 ns |  0.3510 ns |  0.3755 ns |     6.930 ns |     6.736 ns |     8.100 ns |      - |         - |
| SequenceEqual_Original_Sequence | 1000 | 1,840.548 ns | 22.7347 ns | 21.2661 ns | 1,835.535 ns | 1,816.222 ns | 1,877.762 ns |      - |      40 B |
| SequenceEqual_Original_Array    | 1000 |   650.695 ns |  3.7116 ns |  3.0994 ns |   649.948 ns |   646.472 ns |   655.441 ns |      - |         - |
| SequenceEqual_Original_List     | 1000 | 1,302.943 ns |  4.4122 ns |  3.9113 ns | 1,302.234 ns | 1,297.050 ns | 1,311.717 ns |      - |      40 B |
| SequenceEqual_Sequence          | 1000 | 1,861.443 ns | 43.0806 ns | 44.2406 ns | 1,839.284 ns | 1,819.619 ns | 1,982.379 ns |      - |      40 B |
| SequenceEqual_Array             | 1000 |    36.563 ns |  0.3177 ns |  0.2653 ns |    36.451 ns |    36.222 ns |    37.124 ns |      - |         - |
| SequenceEqual_List              | 1000 |    36.467 ns |  0.2840 ns |  0.2657 ns |    36.480 ns |    36.092 ns |    36.840 ns |      - |         - |
```

prozolic committed:
Optimize ImmutableArrayExtensions.SequenceEqual with ICollection<T> fast path

xtqqczze commented:
> * ICollection: ~17% faster (123.20ns → 102.126ns), eliminated allocations for some cases

We cannot make this conclusion from the data, as there is high Error/StdDev for the ICollection benchmark. Were other applications running at the same time as the benchmark?

> IEnumerable: Maintained similar performance while preserving compatibility

We didn't delegate to `Enumerable.SequenceEqual` for this case because the existing implementation had better performance and didn't allocate a second enumerator.

prozolic commented:
> We cannot make this conclusion from the data, as there is high Error/StdDev for the ICollection benchmark. Were other applications running at the same time as the benchmark?

You're right about the ICollection claim. I've updated it to "No statistically significant performance difference" (please let me know if there's a better wording). Regarding benchmark results, other applications may have been running and affecting the measurements, so I re-ran the benchmarks in a clean environment and updated the results.

> We didn't delegate to Enumerable.SequenceEqual for this case because the existing implementation had better performance and didn't allocate a second enumerator.

I've updated it to "Uses existing implementation for this case because the existing implementation had better performance and didn't allocate a second enumerator"

prozolic committed:
Update src/libraries/System.Collections.Immutable/src/System/Linq/ImmutableArrayExtensions.cs

Co-authored-by: xtqqczze <45661989+xtqqczze@users.noreply.github.com>

xtqqczze commented on src/libraries/System.Collections.Immutable/src/System/Linq/ImmutableArrayExtensions.cs:
```diff
@@ -187,28 +187,39 @@ public static bool SequenceEqual<TDerived, TBase>(this ImmutableArray<TBase> imm
-            int i = 0;
-            int n = immutableArray.Length;
-            foreach (TDerived item in items)
+            static bool Impl(ImmutableArray<TBase> immutableArray, IEnumerable<TDerived> items, IEqualityComparer<TBase>? comparer)
```
There may be a difference on R2R due to boxing, I'm not sure.

prozolic replied:
I'm also not very familiar with R2R (ReadyToRun), so I'm sorry but I don't know the impact of this change either.

xtqqczze replied:
> I think you can change the type of the `items` parameter here to `IEnumerable<TBase>` and cast at the call site. Let's see what other reviewers think.

I did an experiment, not sure what conclusions to draw: https://github.com/MihuBot/runtime-utils/issues/1522

prozolic committed:
Merge branch 'main' into ImmutableArrayExtensions.SequenceEqual

pavelsavara committed:
Merge branch 'main' into ImmutableArrayExtensions.SequenceEqual

prozolic committed:
Merge branch 'main' into ImmutableArrayExtensions.SequenceEqual

xtqqczze commented on src/libraries/System.Collections.Immutable/src/System/Linq/ImmutableArrayExtensions.cs:
```diff
@@ -187,28 +187,39 @@ public static bool SequenceEqual<TDerived, TBase>(this ImmutableArray<TBase> imm
+            if (items is ICollection<TBase> itemsCol)
+            {
+                immutableArray.ThrowNullRefIfNotInitialized();
+                return Enumerable.SequenceEqual(immutableArray.array!, itemsCol, comparer);
```
```suggestion
                return Enumerable.SequenceEqual(immutableArray.array, itemsCol, comparer);
```

Not necessary after https://github.com/dotnet/runtime/pull/119898

xtqqczze commented on src/libraries/System.Collections.Immutable/src/System/Linq/ImmutableArrayExtensions.cs:
```diff
@@ -187,28 +187,39 @@ public static bool SequenceEqual<TDerived, TBase>(this ImmutableArray<TBase> imm
+            }
 
-            comparer ??= EqualityComparer<TBase>.Default;
+            return Impl(immutableArray, items, comparer);
```
If you're making more changes, may as well change the method name to `Enumerate` to be more descriptive.

prozolic committed:
Merge branch 'main' into ImmutableArrayExtensions.SequenceEqual

prozolic committed:
Update src/libraries/System.Collections.Immutable/src/System/Linq/ImmutableArrayExtensions.cs

Co-authored-by: xtqqczze <45661989+xtqqczze@users.noreply.github.com>

prozolic committed:
Rename local method from Impl to Enumerate

xtqqczze commented:
@MihuBot benchmark System.Linq.Tests.Perf_ImmutableArrayExtensions

xtqqczze commented:
@EgorBo Blocked on https://github.com/EgorBot/runtime-utils/issues/504

prozolic committed:
Merge branch 'main' into ImmutableArrayExtensions.SequenceEqual

prozolic commented:
@xtqqczze Thank you for the benchmarks and all the helpful reviews!
From the benchmark results, the improvements look good to me.

Benchmark results (MihuBot/runtime-utils#1565, #1566):
- Array: 82% faster (77.44ns → 14.29ns)
- List: 86% faster (107.26ns → 14.68ns)
- IList: 56% faster (180.47ns → 79.82ns)

Is there anything else needed to move this forward?

PranavSenthilnathan reviewed (commented):
Perf results look good. Could you add/update tests to exercise the non-collection path? Currently the tests in `ImmutableArrayExtensionsTest.cs` only test with `ICollection<TDerived>` typed `items` objects.

prozolic committed:
Add tests for non-collection path in ImmutableArrayExtensions.SequenceEqual

prozolic committed:
Merge branch 'main' into ImmutableArrayExtensions.SequenceEqual

EgorBo commented:
@EgorBot -amd --filter "System.Linq.Tests.Perf_ImmutableArrayExtensions*"

prozolic committed:
Merge branch 'main' into ImmutableArrayExtensions.SequenceEqual

jeffhandley reviewed (approved):
The latest numbers are great. Thanks for exploring all the options suggested along the way, @prozolic.

prozolic commented:
Thank you for the merge! I learned a lot from this PR.
